rusoto/core/src/client_inner.rs

```diff
@@ -23,37 +23,61 @@ impl<P, D> ClientInner<P, D> {
 
 impl<P: ProvideAwsCredentials, D: DispatchSignedRequest> ClientInner<P, D> {
     /// Fetch credentials, sign the request and dispatch it.
-    pub fn sign_and_dispatch<E>(&self, request: SignedRequest) -> SignAndDispatch<P, D, E> {
+    pub fn sign_and_dispatch<F: IntoFuture>(&self, request: SignedRequest, response_handler: fn(HttpResponse) -> F) -> SignAndDispatch<P, D, F> {
         SignAndDispatch {
             credentials_provider: self.credentials_provider.clone(),
             dispatcher: self.dispatcher.clone(),
+            response_handler: response_handler,
             state: SignAndDispatchState::Lazy { request: request },
-            _phan: PhantomData
+            timeout: None
         }
     }
 }
 
-pub struct SignAndDispatch<P: ProvideAwsCredentials, D: DispatchSignedRequest, E> {
+pub trait ClientInnerFuture: Future {
+    fn set_timeout(&mut self, timeout: Duration);
+    fn clear_timeout(&mut self);
+}
+
+pub struct SignAndDispatch<P: ProvideAwsCredentials, D: DispatchSignedRequest, F: IntoFuture> {
     credentials_provider: Arc<P>,
     dispatcher: Arc<D>,
-    state: SignAndDispatchState<P, D>,
-    _phan: PhantomData<E>
+    response_handler: fn(HttpResponse) -> F,
+    state: SignAndDispatchState<P, D, F>,
+    timeout: Option<Duration>,
 }
 
-enum SignAndDispatchState<P: ProvideAwsCredentials, D: DispatchSignedRequest> {
+impl<P, D, F> ClientInnerFuture for SignAndDispatch<P, D, F>
+    where P: ProvideAwsCredentials,
+          D: DispatchSignedRequest,
+          F: IntoFuture,
+          F::Error: From<HttpDispatchError> + From<CredentialsError>
+{
+    fn set_timeout(&mut self, timeout: Duration) {
+        self.timeout = Some(timeout);
+    }
+
+    fn clear_timeout(&mut self) {
+        self.timeout = None;
+    }
+}
+
+enum SignAndDispatchState<P: ProvideAwsCredentials, D: DispatchSignedRequest, F: IntoFuture> {
     Lazy { request: SignedRequest },
     FetchingCredentials { future: P::Future, request: SignedRequest },
     Dispatching { future: D::Future },
+    Handling { future: F::Future },
```

Comment:
It's not really clear to me what this state is by just it's name. Perhaps we could change this to something like "waiting", or "polling"

Reply:
- [ ] Find a better name for `SignAndDispatch::Handling`.